### PR TITLE
Add support for the report-to directive

### PR DESCRIPTION
**Pull request by darakian**
This PR adds basic support for the `report-to` directive. I'm not 100% on the test, but I'm following the format outlined here
https://w3c.github.io/reporting/

### CODE REVIEW

**Comment by jonprindiville** [comment]
Hm, but it's possible I'm  misinterpreting, but this PR looks like it expects the _policy_ to contain the reporting-endpoint-dict. My understanding of the docs is that `Report-To` is the header that contains those dict-like reporting-endpoint definitions, while the `Content-Security-Policy`/`Content-Security-Policy-Report-Only` headers would contain a `report-to` directive that simply _refers_ by name to one of the groups defined in a `Report-To` header.

For example [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to#Examples) give an example where the `Report-To` header defines a `csp-endpoint` group, and then the `Content-Security-Policy` contains a `report-to csp-endpoint` directive.

Similarly, [the W3C docs](https://w3c.github.io/reporting/#examples) contain an example (number 1) where a `Report-To` header defines a group called `endpoint-1` and then that group is subsequently referred to from a `Content-Security-Policy` header and a `Public-Key-Pins` header.

**Comment by darakian** [comment]
Ah interesting. I read the `report-to` docs as always delivering the dict, but re-reading it I think you're right that the dict is delivered once and then referenced. In which case implementing this will be a bit more difficult.

**Comment by darakian** [comment]
Thinking about this a bit more. I think to properly implement `report-to` django-csp would need to be aware of each connection as an object and I'm not sure it is at the moment. 

ie. For the first issuance of the `report-to` directive the entire dict would need to be passed (or at least the relevant endpoints). For subsequent responses to the connection the endpoints can be referenced. A corner case that comes to mind is that I could have different `report-to` endpoints configured for different sections of my website and thus there would be a need to update/append/replace the dict that the user has.

Please correct me if I'm wrong on anything.

**Comment by jonprindiville** [comment]
Well, I'll first say that I've done much more reading about the CSP header than about the Report-To header, but I'm happy to give you my impressions. **I think it would be good to get input from the maintainers here as well, though.**

> I think to properly implement report-to django-csp would need to be aware of each connection as an object [...] For the first issuance of the report-to directive the entire dict would need to be passed [...] For subsequent responses to the connection the endpoints can be referenced

Hm, I'm not so sure about connection tracking. My impression was that one _can_ just deliver the `Report-To` header every time, but maybe it's not strictly necessary to do so: there's a required `max_age` field in there telling clients how long they can cache the reporting endpoints.

For example if I'm a server and you GET some document from me, maybe I come back with a response including the following headers:

```
Report-To: { "group": "endpoint-1", "max_age": 10886400, "endpoints": [...] }
Content-Security-Policy: ...; report-to=endpoint-1
```

The CSP refers to the reporting endpoint by name, the reporting endpoint is defined in Report-To.

I _think_ that for subsequent requests _it would be possible_ for me to omit the Report-To and simply include the `Content-Security-Policy: ...; report-to=endpoint-1` because clients are able to cache my `endpoint-1` for 10886400 seconds.

Whether it's worth your while to try to only include a Report-To every once in a while based on `max_age`... eh, I'm not really so sure.

I'm wondering if the intention is that folks just include `Report-To` on all responses. 🤔

**Comment by darakian** [comment]
> I'm wondering if the intention is that folks just include Report-To on all responses.

I'm also unclear on this point. I think the simplest implementation would be to include the entire report dict on each request. 

The `max_age` and timeout logic seems like it requires connection tracking to me, but perhaps I'm missing something. I honestly haven't read into the spec as well as I should have. My target is MVP :)

For clarity; In my initial PR comment I thought that the `report-to` was a field in the CSP header rather than its own header like the `CSP_REPORT_URI` directive. I'll update my PR to set a new header (once I figure out how to do that).

**Comment by darakian** [comment]
Updated with a first pass.

**Comment by g-k** [comment]
Hi @darakian, what's your use case here? Do you have an example CSP header you're trying to serve?

As @jonprindiville helpfully pointed out `report-to` is a separate header that a CSP header can reference from a `report-to` directive.

Given that the `report-to` API is still in draft and people will need to include `report-uri` for cross browser compatibility (per MDN excerpted below), I think django-csp should support the `report-to` CSP directive for people that want to use it, but not send the `report-to` header (users should do that from their CDN, reverse proxy, or separate django middleware).

@EnTeQuAk :+1:, :-1:, or other thoughts?

Assuming that's sounds OK we can pare this PR down to just add the `report-to` directive for setting a string. 

> Though the report-to directive is intended to replace the deprecated report-uri directive, report-to isn’t supported in most browsers yet. So for compatibility with current browsers while also adding forward compatibility when browsers get report-to support, you can specify both report-uri and report-to:
> 
> Content-Security-Policy: ...; report-uri https://endpoint.com; report-to groupname
> 
> In browsers that support report-to, the report-uri directive will be ignored.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri

**Comment by darakian** [comment]
I don't actually have a use case at the moment. I just wanted to add the capability so that it was available down the line :)

**Comment by g-k** [comment]
> I don't actually have a use case at the moment. I just wanted to add the capability so that it was available down the line :)

Cool, yeah it's good to future proof the library a bit, since I think it'd be easy to add the directive.

**Comment by darakian** [comment]
Yep. Let me know what I should pare down and I'll update the PR.

File: csp/utils.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import copy
 import re
 import warnings
+import json
```
Comment (g-k) [inline csp/utils.py]: We can drop this import.

File: csp/utils.py [hunk 2]
```diff
@@ -114,6 +116,40 @@ def build_policy(config=None, update=None, replace=None, nonce=None):
                       for k, val in policy_parts.items()])
 
 
+def build_report_policy(config=None, update=None, replace=None):
```
Comment (g-k) [inline csp/utils.py]: This could also go in a separate library.

File: docs/configuration.rst [hunk 3]
```diff
@@ -77,6 +77,9 @@ These settings affect the policy in the header. The defaults are in
     Set the ``report-uri`` directive. A tuple or list. Each URI can be a
     full or relative URI. *None*
     Note: This doesn't use default-src as a fall-back.
+``CSP_REPORT_TO``
+    Set the ``report-to`` directive. A dict. *None*
```
Comment (g-k) [inline docs/configuration.rst]: We can update this from `A dict.` to something like `A string referring to a reporting group from a report-to header served separately` and link to the draft standard.

File: csp/utils.py (hunk 1 above)
Comment (darakian) [reply csp/utils.py]: Removed.

File: csp/utils.py (hunk 2 above)
Comment (darakian) [reply csp/utils.py]: I've pulled the function.

File: docs/configuration.rst (hunk 3 above)
Comment (darakian) [reply docs/configuration.rst]: Updated. I've made mention of the examples in section 1.2 here
https://w3c.github.io/reporting/#group

File: csp/tests/test_utils.py [hunk 4]
```diff
@@ -135,6 +135,20 @@ def test_report_uri_lazy():
     policy_eq("default-src 'self'; report-uri /foo", policy)
 
 
+@override_settings(CSP_REPORT_TO='{"group": "some_endpoint", \
+    "max_age": 7, \
+    "endpoints": [\
+    { "url": "https://foo.bar/reports_or_something", "priotity": 1}, \
+    { "url": "https://bar.foo/whatever", "priotity": 2}]}')
```
[the remark this answers is not in the document]
Comment (darakian) [reply csp/tests/test_utils.py]: I've removed the logic around `build_report_policy` and tests are passing. Can you give it a look over a double check it?

File: docs/configuration.rst (hunk 3 above)
Comment (darakian) [reply docs/configuration.rst]: If all is well I can squash the commits down before doing a merge too.

File: csp/tests/test_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (g-k) [reply csp/tests/test_utils.py]: The test looks good, but I think the `report-to` value should just be a reference to the group name and not a JSON dict.

File: csp/middleware.py [hunk 5]
```diff
@@ -59,15 +59,17 @@ def process_response(self, request, response):
         if status_code == http_client.INTERNAL_SERVER_ERROR and settings.DEBUG:
             return response
 
-        header = 'Content-Security-Policy'
+        csp_header = 'Content-Security-Policy'
```
[the remark this answers is not in the document]
Comment (darakian) [reply csp/middleware.py]: Reverted.

**Comment by darakian** [comment]
@g-k for the `report-to` directive in `csp/tests/test_utils.py`. Do you mean that the `CSP_REPORT_TO` value should only be set to a single url? ex. `CSP_REPORT_TO='"https://foo.bar/reports_or_something"'`

**Comment by g-k** [comment]
Examples in the reporting API draft show the group name unquoted:

> And the following headers, which direct CSP and HPKP reports to that group:

> Content-Security-Policy: ...; report-to endpoint-1

> Content-Security-Policy: ...; report-to csp-endpoint

https://w3c.github.io/reporting/#examples

**Comment by darakian** [comment]
Gotcha. I should have checked the docs. Updated.

**Comment by darakian** [comment]
Sure thing. You should have let me squash the commits down for cleanliness though :)

**Pull request merged**